### DISCUSSION TITLE
Auto approved PR's should not perform self mutation

### DISCUSSION THREAD

**Discussion by iliapolo** [projen/projen#1413] [General]
I was starting to wonder about the self mutation functionality in the context of auto approved PR's. 

For the sake of discussion, let's talk about the `upgrade-dependencies` PR's created by `projen`, but this applies to all PR's labeled with the `auto-approve` label.

### Concern

Often, these PR's result in a change to the snapshot files of tests. For example, in [construct-hub](https://github.com/cdklabs/construct-hub), it often changes the hashes of the CDK assets we use in our unit tests.

> Example: https://github.com/cdklabs/construct-hub/pull/657/files#

While modifying asset hashes might be considered relatively safe, nothing prevents the build process from significantly changing the snapshot, it really depends on what kind of changes are being pulled during the dependency upgrade. 

This means that currently, we are modifying test snapshots without even looking at what changed.

### Proposal

The catch-all mitigation for this would be stop performing self-mutation commits on auto approved PR's, and instead, perform the anti-temper check. 

Basically as I see it, every PR will have both `self-mutation` and `anti-temper` jobs **defined** in the workflow, but only one of them will **run**, depending on the conditions of the PR. 

Every PR that doesn't perform self-mutation, should have an anti-temper check. And every PR that doesn't have an anti-temper check, should perform self-mutation to keep the code in sync.

----------------------------

### Let think about this

Removing self mutation from the upgrade PR's may re-introduce a lot of the manual burden we were trying to get rid of in the first place. This is especially true for CDK based projects that use assets.  

In addition, a point can be made that snapshots who are susceptible to significant change in an auto approved PR, should be incorporated into integration tests. Integration test snapshots normally won't undergo self-mutation, and if a diff is detected in them, the build will fail. 

### Note

I'm not suggesting to remove the mutation caused by projen synthesis, this should still remain and be part of the initial PR. (i.e the job creating the automatic PR should also run `projen`)

**Comment by eladb** [comment]
Yes, totally agree. We should add a condition that disables self-mutation when a PR is tagged with the `auto-approve` label.
